### DISCUSSION TITLE
Which kind of "update scheduler" behavior make more sense?

### DISCUSSION THREAD

**Discussion by STRRL** [chaos-mesh/chaos-mesh#1491] [Ideas]
Hello guys! We'd like to make a discussion about the behavior when a user updates the scheduler.

When we talk about "cron: @every 10m" with "duration: 5m", I think users do not care about when exactly the chaos starts, users only want to make sure that indeed happens chaos in "each 10 min", and "takes for 5 min".

As description in #1481 says:

> Dynamically modify scheduler field. After reviewing the code in twophase controller, I thought it's hard to define a easy-to-understand behavior when modifying the scheduler field. Maybe we should use webhook to avoid modifying scheduler 💢

Currently, I have collected 3 different strategies of behavior:

1. Update and Retime Immediately
1. Update Since Next Schedule
1. Schedule Independently

I will make some examples to describe this behavior:

1. create one chaos at `00:00`, with `cron: @every 10m` and `duration: 5m`. Then update it at `00:08`, to `cron: @every 20m` and `duration: 10m`. When updating happens, chaos experiments status is `Waiting`.
1. create one chaos at `00:00`, with `cron: @every 10m` and `duration: 5m`. Then update it at `00:13`, to `cron: @every 20m` and `duration: 10m`. When updating happens, chaos experiments status is `Running`.

#### Strategies 1 Update and Retime Immediately

It will apply the new `cron` and `duration` as soon as the CRD updates.

example 1:

![](https://i.imgur.com/kyKjw8n.jpg)

example 2:
![](https://i.imgur.com/PoEiWLR.jpg)

Obversilly, this behavior depends on the update time, it's not kind of stable.

#### Strategies 2 Update Since Next Schedule

It is the current implementation. It do not makes changes on current schedule, but apply new rules since next schedule:

example 1:
![](https://i.imgur.com/q6p1q8w.jpg)

example 2:
![](https://i.imgur.com/yz1VKZZ.jpg)

The behaviors between "update during Running" and "Update during Waiting" might be different. That's my concern.

#### Strategies 3 Schedule Independently

It doesn't care about the previous phase or status, it just calculates the desired status for given `cron`, `duration` and the creation time, then make changes to the desired status. I prefer this strategy.

example 1:
![](https://i.imgur.com/ca35fTQ.jpg)

example 2:
![](https://i.imgur.com/gVL8nVk.jpg)

It just like the new rule already runs concurrently, and you just "switch" to the new time line.
![](https://i.imgur.com/HaetvLF.jpg)

What do you think about it? Welcom new comments && suggestions!

**Comment by Colstuwjx** [comment]
I also prefer `Strategies 3`. Besides, I think we also need to answer the following question: Should we make the `scheduler` a separate component?

For example, see [Litmus Scheduler Usage](https://docs.litmuschaos.io/docs/scheduling/), it's a little bit different design towards our plan ( CronWorkflow and Workflow to orchestrate the chaos, right? ): 
    * Firstly, it supports concrete time schedule, e.g. `startTime: "2020-05-12T05:47:00Z"`;
    * Secondly, it supports one-time job and repeat job, via `once` and `repeat`;
    * Thirdly, it supports `jobCleanUpPolicy`, each CronJob schedule would leave its job histories (  OwnerReferenced with the CronJOb ) and each job instance stands for an experiment.

Currently, IMO, Chaos Workflow and CronWorkflow take the scheduler as one part of their component, it's okay to me, but I think we may also need to think about how to implement similar features about these three points, for me, I'd like to have a concrete time schedule with repeat, e.g. `start at 2020-05-12T05:47:00Z, and keep 30 mins, and fade out soon`.

Thanks.
